network-config-analyzer/CalicoPolicyYamlParser.py

```diff
@@ -45,46 +45,161 @@ def _parse_selector_expr(self, expr, origin_map, namespace=None, is_namespace_se
         """
         expr = expr.strip()  # remove leading and trailing spaces
         expr = ' '.join(expr.split())  # remove redundant spaces
+        if is_namespace_selector and expr == 'global()':
+            return self.peer_container.get_all_global_peers()
+        elif is_namespace_selector:
+            all_peers = self.peer_container.get_all_peers_group(include_globals=False)
+        else:
+            all_peers = self.peer_container.get_namespace_pods(namespace)
         if expr == 'all()':
-            return self.peer_container.get_all_peers_group()
+            return all_peers
 
         # expressions like "has(key)" or "!has(key)"
-        has_re = re.compile('(!?)has\\(([\\w.\\-/]+)\\)')
+        has_re = re.compile('has\\(([\\w.\\-/]+)\\)')
         has_match = has_re.match(expr)
         if has_match:
-            label = has_match.group(2)
+            label = has_match.group(1)
             # TODO: currently misses ANDing of keys
             self.allowed_labels.add(label)
             if is_namespace_selector:
-                return self.peer_container.get_namespace_pods_with_key(label, has_match.group(1) == '!')
-            return self.peer_container.get_peers_with_key(namespace, label, has_match.group(1) == '!')
+                return all_peers & self.peer_container.get_namespace_pods_with_key(label, False)
+            return all_peers & self.peer_container.get_peers_with_key(namespace, label, False)
 
         # expressions like "key==val" or "key!=val"
         key_val_re = re.compile("([\\w.\\-/]+)\\s?([=!])=\\s?'([\\w.\\-/]+)'")
         key_val_match = key_val_re.match(expr)
         if key_val_match:
             key = key_val_match.group(1)
             val = key_val_match.group(3)
+            action = PeerContainer.FilterActionType.In
+            if key_val_match.group(2) == '!':
+                action = PeerContainer.FilterActionType.NotIn
             self.allowed_labels.add(key)
             if is_namespace_selector:
-                return self.peer_container.get_namespace_pods_with_label(key, [val], key_val_match.group(1) == '!')
-            return self.peer_container.get_peers_with_label(key, [val], key_val_match.group(2) == '!', namespace)
+                return all_peers & self.peer_container.get_namespace_pods_with_label(key, [val], action)
+            return all_peers & self.peer_container.get_peers_with_label(key, [val], action)
 
         # expressions like key in "key in {val1, val2}" or "key not in {val1, val2}"
         in_re = re.compile("([\\w.\\-/]+)\\s(not\\s)?in\\s?{\\s?'[\\w.\\-/]+'\\s?(,\\s?'[\\w.\\-/]+'\\s?)*}")
         in_match = in_re.match(expr)
         if in_match:
             key = in_match.group(1)
             values = re.findall("'([\\w.\\-/]*)'", expr)
+            action = PeerContainer.FilterActionType.In
+            if in_match.group(2) is not None:
+                action = PeerContainer.FilterActionType.NotIn
+            if is_namespace_selector:
```

Comment:
To avoid the fw-rules test failure, we can add the following line:
```suggestion
            self.allowed_labels.add(key)
            if is_namespace_selector:
```
It has also revealed a bug in FWRule.py, but I can fix it in another branch.